beets/plugins.py

```diff
@@ -686,8 +694,17 @@ def get_artist(artists, id_key='id', name_key='name'):
             name = re.sub(r' \(\d+\)$', '', name)
             # Move articles to the front.
             name = re.sub(r'^(.*?), (a|an|the)$', r'\2 \1', name, flags=re.I)
+            # Use a join keyword if requested and available.
+            if join_key and artist.get(join_key, None):
+                name += " " + artist[join_key]
+                joined = True
             artist_names.append(name)
-        artist = ', '.join(artist_names).replace(' ,', ',') or None
+        # Concat using spaces only if join_key was passed and the join_key was
+        # ever found. Otherwise join comma-separated.
+        if join_key and joined is True:
+            artist = ' '.join(artist_names) or None
+        else:
+            artist = ', '.join(artist_names).replace(' ,', ',') or None
```

Comment:
I may have this wrong, but would it work to move this "fallback" up into the loop? That is, the loop could look like this:

```py
if join_key and artist.get(join_key, None):
    ...
else:
   name += ', '
```

I'm honestly not 100% sure where the `replace` call should go… it's a little mysterious to me why a ` ,` would appear in the string in the first place.

The reason I suggest this is that it could deal better with a situation where _some_ artists have a `join_key` value and others don't (when there are more than 2 artists). In that case, you'd probably want to use a comma for pairs of artists that would otherwise not have another joiner.

Reply:
Thanks a lot for the suggestion. Well, I think I managed to massively simplify this, which I think makes this method much more readable. It works well with Discogs as the metadata source, but some more testing reveals that it's problematic with the beatport4 plugin. Here we don't have a simple list to handle but a generator which does not answer well on my len() call:

```
  File "/home/jojo/git/beets-beatport4/beetsplug/beatport4.py", line 629, in _get_artist
    return MetadataSourcePlugin.get_artist(
  File "/home/jojo/git/beets/beets/plugins.py", line 697, in get_artist
    if idx < (len(artists) - 1):  # Skip joining on last.
TypeError: object of type 'generator' has no len()
```

Reply:
I found a way to count the items in a way generators as well as dicts and lists or any other iterable should be allowed to be passed as the `artists` object.

Still this is better and more readably than my first version? What do you think?

And btw. I forgot to mention before: I removed the "mysterious" string replace method. It was mysterious to me as well why this was there in the original get_artist() code. I would like to leave it out, I don't see the point of keeping it. It probably would just lead to confusion again in the future :-)